src/jacobi.c

```diff
@@ -0,0 +1,226 @@
+*/
+void jacobiCreateRotationMatrix(real_number* input, vec_size row, vec_size col, real_number* output, vec_size size) {
+
+    real_number aqq = input[coordToIndex(col, col, size)];
+    real_number app = input[coordToIndex(row, row, size)];
+    real_number apq = input[coordToIndex(row, col, size)];
+
+    real_number tau = (aqq - app) / (2.0 * apq);
+    real_number t1 = -tau - sqrt(1 + tau * tau);
+    real_number t2 = -tau + sqrt(1 + tau * tau);
+    real_number t = fabs(t1) > fabs(t2) ? t1 : t2;
+
+    real_number c = 1.0 / (sqrt(1 + t * t));
+    real_number s = c * t;
+
+    vec_size cIndex1 = coordToIndex(row, row, size);
+    vec_size cIndex2 = coordToIndex(col, col, size);
+    vec_size sIndex1 = coordToIndex(row, col, size);
+    vec_size sIndex2 = coordToIndex(col, row, size);
+    vec_size sizeSquared = size * size;
+
+    for (vec_size i = 0; i < sizeSquared; ++i) {
+        if (i == cIndex1 || i == cIndex2) {  
+            output[i] = c;
+        } else if (i == sIndex1) {
+            output[i] = s;
+        } else if (i == sIndex2) {
+            output[i] = -s;
+        } else if (isDiagonal(i, size)) {
+            output[i] = 1;
+        } else {
+            output[i] = 0; // Off diagonal
+        }
+    }
+}
+
+
+/**
+ * Creates the identity matrix (1 on the diagonal and 0 on other coordinates)
+ * @param size size of the matrix
+ * @param output output matrix
+*/
+void jacobiCreateIdentityMatrix(vec_size size, real_number* output) {
+    vec_size sizeSquared = size * size;
+    for (vec_size i = 0; i < sizeSquared; ++i) {
+        if (isDiagonal(i, size)) {
+            output[i] = 1;
+        } else {
+            output[i] = 0;
+        }
+    }
+}
+
+/**
+ * Computes the off diagonal absolute sum of a matrix (sums every element not in the diagonal)
+ * @param matrix matrix to compute the sum on
+ * @param size size of matrix
+*/
+real_number jacobiComputeOffDiagonalSum(real_number* matrix, vec_size size) {
+    real_number sum = 0;
+    vec_size sizeSquared = size * size;
+    
+    for (vec_size i = 0; i < sizeSquared; ++i) {
+        if (!isDiagonal(i, size)) {
+            sum += fabs(matrix[i]);
+        }
+    }
+    return sum;
+}
+
+/**
+ * Computes the jacobi method to find the eigenvalues and eigenvectors of the input matrix
+ * @param inputMatrix matrix whose eigenvalues and eigenvectors we are looking for (will contain the eigenvalues after this function has been called)
+ * @param size size of the matrix (matrix must be square)
+ * @param outputMatrix matrix containing every eigenvector of the input matrix (same size as the input matrix)
+ * @param max_iterations max number of iterations the algorithm can do (set to -1 to allow any amount of iterations)
+ * @param cyclic whether to use the cyclic jacobi method or the classical version
+*/
+void jacobi(real_number* inputMatrix, vec_size size, real_number* outputMatrix, vec_size max_iterations, vec_size cyclic) {
+
+    if (size == 1) {
```

Comment:
Do we really need to handle the case where the size is one? This looks like wasted cycles to me. What do you think ?

Reply:
True